design/service-library.md

```diff
@@ -0,0 +1,187 @@
+
+Models are also available in this package and should reflect the underlying version of Graph we are targeting. Developers using our types should do it in a way that will be familiar to how they are used to with other API Clients and APIs. We should provide an easy-to-use model that feels natural in the JavaScript world. This means we should not be forcing developers to use Classes and / or other structures that will make their code less natural to them.
+
+Because of the richness of the Graph APIs, we are making a decision to require developers to selectively import the areas of Graph they care about. This will provide a lightweight bundle size while still offering a good developer experience.
+
+### Selective Imports of the Graph areas needed
+
+To offer a bundle size that fair to the developer, we are using a selective imports strategy. This strategy enables developers to bring only the endpoints they need from Graph. Developers would only see the root paths of Graph when using the API Client by default, and would need to import additional path with selective imports.
+
+#### Using the service library without any selective imports
+
+> ❌ **Fails** <br />
+> This example does not work because `usersById().directReports` is not available at the root of the Graph Client
+
+```typescript
+const directReports = await graphClient.usersById("ab677c1b-50c6-4013-aa9f-792ed42c59c8").directReports.get();
+```
+
+> ✅ **Succeeds** <br />
+> This example succeeds because `me` is at the root of the Graph Client
+
+```typescript
+const me = await graphClient.me.get();
+```
+> ✅ **Succeeds** <br />
+> This example succeeds because `users` is at the root of the Graph Client
+
+```typescript
+const users = await graphClient.users.get();
+```
+
+#### Using the service library with selective imports
+
+> ❌ **Fails** <br />
+> This example fails because `@microsoft/microsoft-graph-client/users/item` is not imported and therefore not available on the Graph Client
+
+```typescript
+import "@microsoft/microsoft-graph-client/users";
+const directReports = await graphClient.usersById("ab677c1b-50c6-4013-aa9f-792ed42c59c8").directReports.get();
+```
+
+> ✅ **Succeeds** <br />
+> This example succeeds because `/users/item` is imported, making it available on the Graph Client
+
+```typescript
+import "@microsoft/microsoft-graph-client/users/item";
+const directReports = await graphClient.usersById("ab677c1b-50c6-4013-aa9f-792ed42c59c8").directReports.get();
+const extensions = await graphClient.usersById("ab677c1b-50c6-4013-aa9f-792ed42c59c8").extensions.get();
+const extensions = await graphClient.usersById("ab677c1b-50c6-4013-aa9f-792ed42c59c8").extensionsById("85147e97-ed89-4799-b5b8-8d35a9204604").get();
+const joinedTeams = await graphClient.me.joinedTeams.get();
```

Comment:
same for joined teams

Reply:
If I understand correctly Kiota has no way of telling that /me and /users/id are aliases, so there needs to be separate imports here.

Reply:
the repetitive nodes aspects is something we've been talking about while investigating the go build performance issues.
Effectively the OpenAPI spec has no way of describing "this node is a clone of that other node". We could get around that by using an extension.
Now the next problem becomes, does the conversion process make the information available? Semantically speaking, because of OData notations, two nodes can be really similar but not identical (ignoring for description and path). For example, a node could support a PUT/PATCH operation in one context but not in the other, or have a OData method attached in only one context.

Assuming that the conversion library solves for that, we now have a deterministic behaviour issue.
E.g. I convert my metadata and generate SDKs in September, at this time /me and everything under it is identical to /users/id and everything under it. 
The SDK generates a single set of request builders, inline requests and responses types. 
Then in October somebody adds an annotation on /me but not on /users/id. The conversion process judges that those nodes are now different, produces the description accordingly, and Kiota generates two sets of types. **Effectively breaking half the applications** that had references "from the wrong path".
We could major bump every-time that happens, but with a service the size of Microsoft Graph and the highly repetitive nature of OData I anticipate this is going to be extremely painful for us and for our customers.

So the conclusion of that was not to deduplicate nodes automatically but make conscious and careful decisions when we see opportunities like that using kiota's filtering capability and a little bit of handcrafted code in the service lib. A good example of that is /me/ and /users/id. They overlap 100% today, are huge, and even if they started drifting in the future, we'd keep the deduplication going and tell customers impacted by the edge case that's it's a trade off decision for the size of the SDK.

You can see an example of that [here](https://github.com/microsoftgraph/msgraph-sdk-go/pull/333), it's only a couple of minutes of work, yields about a 17% reduction in build time (and probably similar in size, but I haven't measured that).

Reply:
The additional aspect I forgot to mention is that we could simplify the selective import story in TypeScript by manually maintaining "presets", a little like pnpJS does. Where we'd have on one hand the granular modular augmentation (each node level) generated by kiota, and on the other presets that leverage them.
And examples would be

```TypeScript
import "msgraphsdk/presets/user";
import "msgraphsdk/presets/teams";

await client.me.joinedTeams.get();
await client.users.byId("foo").joinedTeams.get();
```

In that example both me and users are here because of the user preset, and joined teams is available under both of them because of the teams preset.